File: java/src/test/java/codingdojo/AuctionMessageTranslatorApprovalTest.java
package codingdojo;

import org.junit.jupiter.api.Test;

import static org.approvaltests.Approvals.verify;


public class AuctionMessageTranslatorApprovalTest {
    @Test
    public void notifiesAuctionClosedWhenCloseMessageReceived() {
        String message = "SOLVersion: 1.1; Event: CLOSE;";
        var listener = new AuctionEventListenerApprovalSpy();
        var translator = new AuctionMessageTranslator(listener);

        translator.processMessage(message);

        verify(listener);
    }

    @Test
    public void notifiesBidDetailsWhenPriceMessageReceived() {
        String message = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 192; Increment: 7; Bidder: Someone else;";
        var listener = new AuctionEventListenerApprovalSpy();
        var translator = new AuctionMessageTranslator(listener);

        translator.processMessage(message);

        verify(listener);
    }

    static class AuctionEventListenerApprovalSpy implements AuctionEventListener {

        String log = "";
        @Override
        public void auctionClosed() {
            log += "auctionClosed";
        }

        @Override
        public void currentPrice(int price) {
            log += "currentPrice " + price;

        }

        @Override
        public void auctionFailed() {
            log += "auctionFailed";
        }

        @Override
        public String toString() {
            return log;
        }
    }
}

File: kotlin/src/test/kotlin/org/sammancoaching/AuctionMessageTranslatorApprovalTest.kt
package org.sammancoaching

import org.approvaltests.Approvals
import kotlin.test.Test

class AuctionMessageTranslatorApprovalTest {
    @Test
    fun notifiesAuctionClosedWhenCloseMessageReceived() {
        val spy = StringBuilder()
        val message = "SOLVersion: 1.1; Event: CLOSE;"
        spy.append("Processing event: $message\n")
        val listener = AuctionEventListenerApprovalSpy(spy)
        val translator = AuctionMessageTranslator(listener)

        translator.processMessage(message)

        Approvals.verify(spy)
    }

    @Test
    fun notifiesBidDetailsWhenPriceMessageReceived() {
        val spy = StringBuilder()
        val message = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 192; Increment: 7; Bidder: Someone else;"
        spy.append("Processing event: $message\n")
        val listener = AuctionEventListenerApprovalSpy(spy)
        val translator = AuctionMessageTranslator(listener)

        translator.processMessage(message)

        Approvals.verify(spy)
    }
}